darkcluster/src/main/java/com/linkedin/darkcluster/impl/DarkClusterStrategyFactoryImpl.java

```diff
@@ -0,0 +1,118 @@
+package com.linkedin.darkcluster.impl;
+
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nonnull;
+
+import com.linkedin.common.util.Notifier;
+import com.linkedin.d2.DarkClusterConfig;
+import com.linkedin.d2.DarkClusterConfigMap;
+import com.linkedin.d2.balancer.LoadBalancerClusterListener;
+import com.linkedin.d2.balancer.ServiceUnavailableException;
+import com.linkedin.d2.balancer.util.ClusterInfoProvider;
+import com.linkedin.darkcluster.api.BaseDarkClusterDispatcher;
+import com.linkedin.darkcluster.api.DarkClusterDispatcher;
+import com.linkedin.darkcluster.api.DarkClusterStrategy;
+import com.linkedin.darkcluster.api.DarkClusterStrategyFactory;
+import com.linkedin.darkcluster.api.DarkClusterVerifierManager;
+import com.linkedin.darkcluster.api.NoOpDarkClusterStrategy;
+
+/**
+ * DarkClusterStrategyFactoryImpl creates and maintains the strategies needed for dark clusters. This involves refreshing
+ * when darkClusterConfig changes are detected, by way of a {@link LoadBalancerClusterListener}
+ */
+public class DarkClusterStrategyFactoryImpl implements DarkClusterStrategyFactory
+{
+  private final ClusterInfoProvider _clusterInfoProvider;
+  private final String _sourceClusterName;
+  private final DarkClusterDispatcher _darkClusterDispatcher;
+  private final Notifier _notifier;
+  private final Map<String, AtomicReference<DarkClusterStrategy>> _darkStrategyMap;
+  private final Random _random;
+  private final LoadBalancerClusterListener _clusterListener;
+  private final DarkClusterVerifierManager _verifierManager;
+
+  public DarkClusterStrategyFactoryImpl(@Nonnull ClusterInfoProvider clusterInfoProvider,
+                                        @Nonnull String clusterName,
+                                        @Nonnull DarkClusterDispatcher darkClusterDispatcher,
+                                        @Nonnull Notifier notifier,
+                                        @Nonnull Random random,
+                                        @Nonnull DarkClusterVerifierManager verifierManager)
+  {
+    _clusterInfoProvider = clusterInfoProvider;
+    _sourceClusterName = clusterName;
+    _notifier = notifier;
+    _darkStrategyMap = new ConcurrentHashMap<>();
+    _random = random;
+    _darkClusterDispatcher = darkClusterDispatcher;
+    _verifierManager = verifierManager;
+    _clusterListener = new LoadBalancerClusterListener()
+    {
+      @Override
+      public void onClusterAdded(String updatedClusterName)
+      {
+        if (_darkStrategyMap.containsKey(updatedClusterName))
+        {
+          // this is a dark cluster name. however, to refresh the strategies, we need to pull the
+          // darkClusterConfigMap on the parent d2 cluster, because that has the properties needed
+          // to recreate the dark cluster strategies, such as the multiplier.
```

Comment:
I added clarifying comments to the code. We actually will be listening to both source cluster (because we call ClusterInfoProvider.getDarkClusters() on the source cluster, which calls listenToCluster) and the dark clusters, because we send requests through the d2 client, which also calls listenToCluster. (we also might be calling getClusterCount as well on both source and target cluster). So while it's possible to update strategies when either the source cluster or dark cluster changes, it's most precise doing it only when we receive onClusterAdded for darkClusters, because we know for sure then that the dark cluster znode changed, rather than some unrelated source cluster change (or a different dark cluster change).

Reply:
I see, thanks for clarifying.

Reply:
closing comment.